services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/doi/DoiManagerImpl.java

```diff
@@ -94,45 +94,87 @@ public Doi createOrUpdateDoi(final Long userId, final Doi dto) throws Recoverabl
+		 */
+		DoiAssociation association = createOrUpdateAssociation(dto);
+		dto.setDoiUri(generateDoiUri(dto.getObjectId(), dto.getObjectType(), dto.getObjectVersion()));
+		dto.setDoiUrl(generateLocationRequestUrl(dto.getObjectId(), dto.getObjectType(), dto.getObjectVersion()));
+		DataciteMetadata metadata = createOrUpdateDataciteMetadata(dto);
+		return mergeMetadataAndAssociation(metadata, association);
+	}
+
+	DoiAssociation createOrUpdateAssociation(DoiAssociation dto) throws RecoverableMessageException {
+		DoiAssociation association;
 		try {
+			// Check if the DOI exists by checking to make sure eTags match
+			// (We will get a NotFoundException if it doesn't exist).
 			if (!doiAssociationDao.getEtagForUpdate(dto.getObjectId(), dto.getObjectType(), dto.getObjectVersion()).equals(dto.getEtag())) {
 				throw new ConflictingUpdateException("eTags do not match.");
 			}
-			doiAssociationDao.updateDoiAssociation(dto);
-		} catch (NotFoundException e1) {
+			association = doiAssociationDao.updateDoiAssociation(dto);
+		} catch (NotFoundException e1) { // The DOI does not already exist (exception was thrown by getEtag)
 			try {
-				doiAssociationDao.createDoiAssociation(dto); //retry on fail
-			} catch (Exception e2) {
+				association = doiAssociationDao.createDoiAssociation(dto); // Create
+			} catch (DatastoreException e2) {
+				/*
+				 * If the create method fails after encountering the NotFoundException, it is possible that a separate
+				 * client was able to create a database entry, resulting in a DataStoreException.
+				 * We suggest that the client retry calling the method, which will either lead to a successful update
+				 * call, or attempt will to create the DOI once more (if another caller failed).
+				 */
 				throw new RecoverableMessageException(e2);
 			}
 		}
+		return association;
+	}
+
+	DataciteMetadata createOrUpdateDataciteMetadata(Doi dto) throws RecoverableMessageException {
+		if (dto.getDoiUri() == null) {
+			throw new IllegalArgumentException("DOI URI cannot be null");
+		}
+		if (dto.getDoiUrl() == null) {
+			throw new IllegalArgumentException("DOI URL cannot be null");
+		}
 
-		String uri = generateDoiUri(dto.getObjectId(), dto.getObjectType(), dto.getObjectVersion());
-		String url = generateLocationRequestUrl(dto.getObjectId(), dto.getObjectType(), dto.getObjectVersion());
 		try {
-			dataciteClient.registerMetadata(dto, uri);
-			dataciteClient.registerDoi(uri, url);
+			dataciteClient.registerMetadata(dto, dto.getDoiUri());
+			dataciteClient.registerDoi(dto.getDoiUri(), dto.getDoiUrl());
 			if (dto.getStatus() == DataciteRegistrationStatus.REGISTERED) {
-				dataciteClient.deactivate(uri);
+				dataciteClient.deactivate(dto.getDoiUri());
 			}
+			return dataciteClient.get(dto.getDoiUri());
 		} catch (NotReadyException | ServiceUnavailableException e) {
+			/*
+			 * If the client to the external receives one of these exceptions, the external API may be temporarily down.
```

Comment:
The second call to datacite might file since the calls are 'eventually consistent'.  For either case we retry.